src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5385,6 +5385,471 @@ address generate_avx_ghash_processBlocks() {
+
+    // This function will always write 128 bytes into the key stream buffer
+    // and that length should be returned through %rax.
+    __ mov64(rax, 128);
+
+    __ leave();
+    __ ret(0);
+    return start;
+  }
+
+  /* Add mask for 4-block ChaCha20 Block calculations */
+  address chacha20_ctradd_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20_ctradd_avx512");
+    address start = __ pc();
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000001, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000002, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    __ emit_data64(0x0000000000000003, relocInfo::none);
+    __ emit_data64(0x0000000000000000, relocInfo::none);
+    return start;
+  }
+
+  /* Scatter mask for key stream output on AVX-512 */
+  address chacha20_scmask_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20_scmask_avx512");
+    address start = __ pc();
+    __ emit_data64(0x0000000100000000, relocInfo::none);
+    __ emit_data64(0x0000000300000002, relocInfo::none);
+    __ emit_data64(0x0000001100000010, relocInfo::none);
+    __ emit_data64(0x0000001300000012, relocInfo::none);
+    __ emit_data64(0x0000002100000020, relocInfo::none);
+    __ emit_data64(0x0000002300000022, relocInfo::none);
+    __ emit_data64(0x0000003100000030, relocInfo::none);
+    __ emit_data64(0x0000003300000032, relocInfo::none);
+    return start;
+  }
+
+  /* Scatter mask for collating key stream output in AVX-512 */
+  address generate_chacha20Block_avx512() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "chacha20Block");
+    address start = __ pc();
+
+    Label L_twoRounds;
+    const Register state        = c_rarg0;
+    const Register result       = c_rarg1;
+    const Register loopCounter  = rcx;
+    const KRegister writeMask = k1;
+
+    const XMMRegister zmm_aVec = xmm0;
+    const XMMRegister zmm_bVec = xmm1;
+    const XMMRegister zmm_cVec = xmm2;
+    const XMMRegister zmm_dVec = xmm3;
+    const XMMRegister zmm_scratch = xmm4;
+
+    __ enter();
+
+    // Load the initial state in columnar orientation
+    // We will broadcast each 128-bit segment of the state array into
+    // all four double-quadword slots on ZMM registers.
+    __ evbroadcasti32x4(zmm_aVec, Address(state, 0), Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_bVec, Address(state, 16), Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_cVec, Address(state, 32), Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_dVec, Address(state, 48), Assembler::AVX_512bit);
+
+    __ vpaddd(zmm_dVec, zmm_dVec, ExternalAddress(StubRoutines::x86::chacha20_counter_addmask_avx512()), Assembler::AVX_512bit, rax);
```

Comment:
The chacha20_counter_addmask_avx512() could be preloaded into zmm_s5 before line 5735 as follows:
 __ evmovdquq(zmm_s5, ExternalAddress(StubRoutines::x86::chacha20_counter_addmask_avx512()), Assembler::AVX_512bit, rax);
vpaddd can then use zmm_s5 also the later usage could use zmm_s5 directly.

Reply:
Another good improvement, done.